library/alloc/src/collections/vec_deque/iter.rs

```diff
@@ -13,76 +10,72 @@ use super::{count, wrap_index, RingSlices};
 /// [`iter`]: super::VecDeque::iter
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct Iter<'a, T: 'a> {
-    ring: &'a [MaybeUninit<T>],
-    tail: usize,
-    head: usize,
+    i1: slice::Iter<'a, T>,
+    i2: slice::Iter<'a, T>,
 }
 
 impl<'a, T> Iter<'a, T> {
-    pub(super) fn new(ring: &'a [MaybeUninit<T>], tail: usize, head: usize) -> Self {
-        Iter { ring, tail, head }
+    pub(super) fn new(i1: slice::Iter<'a, T>, i2: slice::Iter<'a, T>) -> Self {
+        Self { i1, i2 }
     }
 }
 
 #[stable(feature = "collection_debug", since = "1.17.0")]
 impl<T: fmt::Debug> fmt::Debug for Iter<'_, T> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let (front, back) = RingSlices::ring_slices(self.ring, self.head, self.tail);
-        // Safety:
-        // - `self.head` and `self.tail` in a ring buffer are always valid indices.
-        // - `RingSlices::ring_slices` guarantees that the slices split according to `self.head` and `self.tail` are initialized.
-        unsafe {
-            f.debug_tuple("Iter")
-                .field(&MaybeUninit::slice_assume_init_ref(front))
-                .field(&MaybeUninit::slice_assume_init_ref(back))
-                .finish()
-        }
+        f.debug_tuple("Iter").field(&self.i1.as_slice()).field(&self.i2.as_slice()).finish()
     }
 }
 
 // FIXME(#26925) Remove in favor of `#[derive(Clone)]`
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T> Clone for Iter<'_, T> {
     fn clone(&self) -> Self {
-        Iter { ring: self.ring, tail: self.tail, head: self.head }
+        Iter { i1: self.i1.clone(), i2: self.i2.clone() }
```

Comment:
What's your plan for the iterator implementation long-term?

If it's going to stay as two iterators and safe code, then this should just be `derive(Clone)` since that's all this is doing.  (And then maybe it should delegate to `Chain`, which already has all the implementations for "use these two iterators".)

If the plan is to have a subsequent PR that makes these iterators smaller by going back to having a buffer and two indices, then please drop a `// FIXME` to that effect in it somewhere.

Reply:
I wasn't planning on making the type any smaller than this, I don't even think that's possible without having to do double indirections on each `next()` call (you need at least the buffer, its capacity, the current index and the number of remaining items). 

`derive(Clone)` would result in an `impl<T> Clone for Iter<'_, T> where T: Clone` wouldn't it? 

And about using `Chain`, because `Chain` needs to be able to handle 2 different iterator types, it wouldn't be able to always keep the first iterator non-empty, which would make it so that in the entire second slice of the iterator, the hot path couldn't be taken in `next()` calls. Looking at the `Chain` source, it also wraps both its iterators in `Option`s, which could lose a potential niche optimization and also potentially further degrade performance. I suppose I could try to benchmark `Chain` against my manual impl, but I'd rather just keep it the way it is right now.

Also, if we ever want to add the methods I talked about on Zulip (`Iter(Mut)::as_(mut_)slices` and `IterMut::into_mut_slices`), then this implementation would make that way easier than if we were using `Chain`.

Reply:
Oh yeah, the generic bounds 🤦